### PR TITLE
handling owner changing

### PR DESCRIPTION
**Pull request by AlenaShalamberydze**

### CODE REVIEW

File: pets-app-sap-cap/srv/cat-service.cds [hunk 1]
```diff
@@ -1,15 +1,20 @@
 using my.application as my from '../db/data-model';
```
Comment (ilyaaleks) [inline pets-app-sap-cap/srv/cat-service.cds:1]: Also don't forget define a namespace

File: pets-app-sap-cap/srv/src/main/java/com/leverx/pets/handlers/MessageKeys.java [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package com.leverx.pets.handlers;
+
```
Comment (ilyaaleks) [inline pets-app-sap-cap/srv/src/main/java/com/leverx/pets/handlers/MessageKeys.java:2]: private noArgsConstructor

File: pets-app-sap-cap/srv/src/main/java/com/leverx/pets/handlers/CatalogServiceHandler.java [hunk 3]
```diff
@@ -1,13 +1,169 @@
...
+        this.analyzer = CqnAnalyzer.create(model);
+        this.messages = messages;
+    }
+
+    @Before(entity = Animals_.CDS_NAME)
+    public void beforeChangeOwner(ChangeOwnerContext context) {
+        log.info("beforeChangeOwner method started...");
+
+        String newOwnerId = context.getNewOwnerID();
+        checkUserExistenceById(newOwnerId);
```
Comment (ilyaaleks) [inline pets-app-sap-cap/srv/src/main/java/com/leverx/pets/handlers/CatalogServiceHandler.java:60]: Move this logic to another class, according to the principle of solid

File: pets-app-sap-cap/srv/src/main/java/com/leverx/pets/handlers/CatalogServiceHandler.java [hunk 4]
```diff
@@ -1,13 +1,169 @@
...
+
+        String animalId = (String) analyzer.analyze(context.getCqn()).targetKeys().get(ID);
```
Comment (ilyaaleks) [inline pets-app-sap-cap/srv/src/main/java/com/leverx/pets/handlers/CatalogServiceHandler.java:62]: Also create utility class for this logic

File: pets-app-sap-cap/srv/src/main/java/com/leverx/pets/handlers/CatalogServiceHandler.java [hunk 5]
```diff
@@ -1,13 +1,169 @@
...
+        Animals animal = selectAnimalByIdIfExists(animalId);
+
+        if (newOwnerId.equals(animal.getUserId())) {
+            log.error("Troubles: User owns this animal already!");
+            throw new ServiceException(NOT_ACCEPTABLE, USER_IS_OWNER);
+        }
```
Comment (ilyaaleks) [inline pets-app-sap-cap/srv/src/main/java/com/leverx/pets/handlers/CatalogServiceHandler.java:68]: Also you could make a Validator Class

File: pets-app-sap-cap/srv/src/main/java/com/leverx/pets/handlers/CatalogServiceHandler.java [hunk 6]
```diff
@@ -1,13 +1,169 @@
...
+    }
+
+    @On(entity = Animals_.CDS_NAME)
+    public void onChangeOwner(ChangeOwnerContext context) {
+        log.info("onChangeOwner method started...");
+        String newOwnerId = context.getNewOwnerID();
+        String animalId = (String) analyzer.analyze(context.getCqn()).targetKeys().get(ID);
```
Comment (ilyaaleks) [inline pets-app-sap-cap/srv/src/main/java/com/leverx/pets/handlers/CatalogServiceHandler.java:75]: Replace with a method of the created class

File: pets-app-sap-cap/srv/src/main/java/com/leverx/pets/handlers/CatalogServiceHandler.java [hunk 7]
```diff
@@ -1,13 +1,169 @@
...
+        Animals animalToUpdate = db
+                .run(Select.from(ANIMALS)
+                        .where(animal -> animal.ID().eq(animalId)))
+                .single(Animals.class);
+
+        animalToUpdate.setUserId(newOwnerId);
+
+        Animals updatedAnimal = db
+                .run(Update.entity(ANIMALS).data(animalToUpdate))
+                .single(Animals.class);
```
Comment (ilyaaleks) [inline pets-app-sap-cap/srv/src/main/java/com/leverx/pets/handlers/CatalogServiceHandler.java:86]: Move the logic of working with the database into a separate repository class

File: pets-app-sap-cap/srv/src/main/java/com/leverx/pets/handlers/CatalogServiceHandler.java [hunk 8]
```diff
@@ -1,13 +1,169 @@
...
+
+    @Before(entity = Users_.CDS_NAME)
+    public void beforeChangeOwnerForSet(ChangeOwnerForSetContext context) {
+        log.info("beforeChangeOwnerForSetContext method started...");
+        String newOwnerId = context.getNewOwnerID();
+        String type = context.getType();
+        String currentOwnerId = (String) analyzer.analyze(context.getCqn()).targetKeys().get(ID);
+
+        checkUserExistenceById(newOwnerId);
+        checkUserExistenceById(currentOwnerId);
```
Comment (ilyaaleks) [inline pets-app-sap-cap/srv/src/main/java/com/leverx/pets/handlers/CatalogServiceHandler.java:100]: Also make a validator class, which will be the spring component. Inject the repository class into this validator and check existence of the user

File: pets-app-sap-cap/srv/src/main/java/com/leverx/pets/handlers/CatalogServiceHandler.java [hunk 9]
```diff
@@ -1,13 +1,169 @@
...
+
+        if (newOwnerId.equals(currentOwnerId)) {
+            log.error("Troubles: User owns this animal already!");
+            throw new ServiceException(NOT_ACCEPTABLE, USER_IS_OWNER);
+        }
```
Comment (ilyaaleks) [inline pets-app-sap-cap/srv/src/main/java/com/leverx/pets/handlers/CatalogServiceHandler.java:105]: Do the same as described above

File: pets-app-sap-cap/srv/src/main/java/com/leverx/pets/handlers/CatalogServiceHandler.java [hunk 10]
```diff
@@ -1,13 +1,169 @@
...
+                        .where(animal -> animal.user_ID().eq(currentOwnerId)));
+
+        List<Animals> updatedAnimals = new ArrayList<>();
+
+        animalsToUpdate.forEach(animal -> {
+            animal.as(Animals.class).setUserId(newOwnerId);
+            updatedAnimals
+                    .add(db.run(Update.entity(ANIMALS).data(animal))
+                            .single(Animals.class));
+        });
```
[the remark this answers is not in the document]
Comment (ilyaaleks) [reply pets-app-sap-cap/srv/src/main/java/com/leverx/pets/handlers/CatalogServiceHandler.java:129]: But don't forget move logic of `bulkUpdateAnimal` method to separate class.

File: pets-app-sap-cap/srv/src/main/java/com/leverx/pets/validators/UserValidator.java [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+import org.springframework.stereotype.Component;
+
+import static com.leverx.pets.util.MessageKeys.USER_NOT_FOUND;
+import static com.sap.cds.services.ErrorStatuses.NOT_FOUND;
+
+@Component
+@Slf4j
+public final class UserValidator {
+
+    private static UserRepository userRepository;
```
Comment (IlyaAliakseyeu) [inline pets-app-sap-cap/srv/src/main/java/com/leverx/pets/validators/UserValidator.java:17]: remove `static`, add `final`.

File: pets-app-sap-cap/srv/src/main/java/com/leverx/pets/validators/UserValidator.java [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+
+    @Autowired
+    public void setUserRepository(UserRepository repo){
```
Comment (IlyaAliakseyeu) [inline pets-app-sap-cap/srv/src/main/java/com/leverx/pets/validators/UserValidator.java:20]: You could use @AllArgsConstructor

File: pets-app-sap-cap/srv/src/main/java/com/leverx/pets/validators/CatalogServiceValidator.java [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+
+    public static void validateChangeOwner(String animalId, String newOwnerId) {
+        checkUserExistenceById(newOwnerId);
+        Animals animal = selectAnimalByIdIfExists(animalId);
+        checkIfUserIsOwner(newOwnerId, animal.getUserId());
+    }
+
+    public static void validateChangeOwnerForSet(String newOwnerId, String currentOwnerId, String animalType) {
+        checkUserExistenceById(newOwnerId);
+        checkUserExistenceById(currentOwnerId);
```
Comment (IlyaAliakseyeu) [inline pets-app-sap-cap/srv/src/main/java/com/leverx/pets/validators/CatalogServiceValidator.java:27]: You could make a method in UserValidator class and use him there:
```
checkUsersExistenceByIds(final String... ids)
{
 Arrays.stream(ids).forEach(VetSchedulingService::checkUserExistenceById);
}
```

File: pets-app-sap-cap/srv/src/main/java/com/leverx/pets/validators/UserValidator.java [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import static com.leverx.pets.util.MessageKeys.USER_NOT_FOUND;
+import static com.sap.cds.services.ErrorStatuses.NOT_FOUND;
+
+@Component
+@Slf4j
+public final class UserValidator {
```
Comment (IlyaAliakseyeu) [inline pets-app-sap-cap/srv/src/main/java/com/leverx/pets/validators/UserValidator.java:15]: The principle of SOLID is violated
https://ru.wikipedia.org/wiki/%D0%9F%D1%80%D0%B8%D0%BD%D1%86%D0%B8%D0%BF_%D0%BE%D1%82%D0%BA%D1%80%D1%8B%D1%82%D0%BE%D1%81%D1%82%D0%B8/%D0%B7%D0%B0%D0%BA%D1%80%D1%8B%D1%82%D0%BE%D1%81%D1%82%D0%B8

File: pets-app-sap-cap/srv/src/main/java/com/leverx/pets/validators/AnimalValidator.java [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+public final class AnimalValidator {
+
+    private static AnimalRepository animalRepository;
+
+    @Autowired
+    public void setAnimalRepository(AnimalRepository repo){
+        animalRepository = repo;
+    }
+
+    public static Animals selectAnimalByIdIfExists(String animalId) {
```
Comment (IlyaAliakseyeu) [inline pets-app-sap-cap/srv/src/main/java/com/leverx/pets/validators/AnimalValidator.java:29]: Make all methods non-static in validators

**Review by IlyaAliakseyeu — APPROVED** [review]

**Pull request merged**